libs/telemetry/logger.go

```diff
@@ -0,0 +1,114 @@
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/databricks/cli/libs/log"
+	"github.com/google/uuid"
+)
+
+// Interface abstraction created to mock out the Databricks client for testing.
+type DatabricksApiClient interface {
+	Do(ctx context.Context, method, path string,
+		headers map[string]string, request, response any,
+		visitors ...func(*http.Request) error) error
+}
+
+func Log(ctx context.Context, event DatabricksCliLog) {
+	l := fromContext(ctx)
+
+	l.logs = append(l.logs, FrontendLog{
+		// The telemetry endpoint deduplicates logs based on the FrontendLogEventID.
+		// This it's important to generate a unique ID for each log event.
+		FrontendLogEventID: uuid.New().String(),
+		Entry: FrontendLogEntry{
+			DatabricksCliLog: event,
+		},
+	})
+}
+
+type logger struct {
+	logs []FrontendLog
+}
+
+// This function is meant to be only to be used in tests to introspect the telemetry logs
+// that have been logged so far.
+func GetLogs(ctx context.Context) []FrontendLog {
+	l := fromContext(ctx)
+	return l.logs
+}
+
+// Maximum additional time to wait for the telemetry event to flush. We expect the flush
+// method to be called when the CLI command is about to exist, so this caps the maximum
+// additional time the user will experience because of us logging CLI telemetry.
+var MaxAdditionalWaitTime = 5 * time.Second
```

Comment:
I'll work on a followup PR to try and reduce the additional timeout we wait to capture telemetry requests.